File: Data Structures & Algorithms/count-number-of-islands/submission-0.java
class Solution {
    public int numIslands(char[][] grid) {
        int result = 0;

        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == '1') {
                    result++;
                    dfs(grid, i, j);
                }
            }
        }

        return result;
    }

    private void dfs(
        char[][] grid,
        int row,
        int col
    ) {
        grid[row][col] = 'a';

        if (row + 1 < grid.length && grid[row + 1][col] == '1') {
            dfs(grid, row + 1, col);
        }

        if (row - 1 > -1 && grid[row - 1][col] == '1') {
            dfs(grid, row - 1, col);
        }

        if (col + 1 < grid[row].length && grid[row][col + 1] == '1') {
            dfs(grid, row, col + 1);
        }

        if (col - 1 > -1 && grid[row][col - 1] == '1') {
            dfs(grid, row, col - 1);
        }
    }
}

File: Data Structures & Algorithms/unique-paths-ii/submission-0.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        row, col = len(obstacleGrid), len(obstacleGrid[0])

        trace = [[0 for _ in range(col)] for _ in range(row)]

        for i in range(col):
            if obstacleGrid[0][i] == 1:
                break
            trace[0][i] = 1

        for i in range(row):
            if obstacleGrid[i][0] == 1:
                break
            trace[i][0] = 1

        for i in range(1, row):
            for j in range(1, col):
                if obstacleGrid[i][j] == 1:
                    trace[i][j] = 0
                else:
                    trace[i][j] = trace[i - 1][j] + trace[i][j - 1]

        return trace[row - 1][col - 1]